File: examples/code-review/steps/code-review/controller.step.ts
import { z } from 'zod';
import { EventConfig, StepHandler } from 'motia';
import { Commits } from '../shared/utils/repository';
import { evaluateCommits } from '../shared/agents/claude';

const mctsControllerInputSchema = z.object({
  prompt: z.string(),
  repoUrl: z.string(),
  branch: z.string(),
  maxIterations: z.number().default(100),
  explorationConstant: z.number().default(1.414),
  maxDepth: z.number().default(10),
  reviewStartCommit: z.string().optional(),
  reviewEndCommit: z.string().optional(),
  requirements: z.string(),
  outputPath: z.string().optional()
});
export type MCTSControllerInput = z.infer<typeof mctsControllerInputSchema>;

export const config: EventConfig = {
  type: 'event',
  name: 'MCTSController',
  description: 'Controls the MCTS process for code review reasoning',
  subscribes: ['review.requested', 'mcts.backpropagation.completed'],
  emits: ['mcts.iteration.started', 'mcts.iterations.completed', 'review.error'],
  flows: ['code-review-flow'],
  input: mctsControllerInputSchema
};

export const handler: StepHandler<typeof config> = async (input: MCTSControllerInput, { emit, logger, state, traceId }) => {
  console.log('Controller received event with input:', JSON.stringify({
    ...input,
    requirements: input.requirements ? 
      (input.requirements.length > 20 ? `${input.requirements.slice(0, 20)}...` : input.requirements) : undefined,
    repoUrl: input.repoUrl
  }, null, 2));
  
  logger.info('Analyzing review context', {
    ...input,
    requirements: input.requirements && input.requirements.length > 20
      ? `${input.requirements.slice(0, 20)}...`
      : input.requirements,
    repoUrl: input.repoUrl
  });

  // Store requirements in state for error handling
  await state.set(traceId, 'requirements', input.requirements);
 
  try {
    logger.info('Creating commits object for repository', { 
      repoUrl: input.repoUrl, 
      branch: input.branch,
      reviewStartCommit: input.reviewStartCommit,
      reviewEndCommit: input.reviewEndCommit
    });
    
    const commits = await Commits.create(traceId, state, input);
    
    logger.info('Successfully accessed repository and loaded commits', {
      filesChanged: commits.files.split('\n').length,
      commitMessages: commits.messages.split('\n').length
    });
    
    const evaluation = await evaluateCommits(commits, input.prompt);

    // Define a unique root node ID
    const rootId = `root-${Date.now()}`;

    // For the first iteration, initialize the MCTS tree structure
    const nodes: Record<string, any> = {
      [rootId]: {
        id: rootId,
        parent: null,
        children: [],
        visits: 1,
        value: 0,
        state: evaluation.summary,
        isTerminal: false
      }
    };
  
    if (evaluation.score > 0.9 || input.maxIterations === 0) {
      // If the score is already high or no iterations requested, complete immediately
      await emit({
        topic: 'mcts.iterations.completed',
        data: {
          nodes,
          rootId,
          currentIteration: 0,
          maxIterations: input.maxIterations,
          explorationConstant: input.explorationConstant,
          maxDepth: input.maxDepth,
          outputPath: input.outputPath
        }
      });
      logger.info('Context analysis completed without iterations');
    } else {
      // Start the MCTS process
      await emit({
        topic: 'mcts.iteration.started',
        data: {
          nodes,
          rootId,
          currentNodeId: rootId,
          currentIteration: 0,
          maxIterations: input.maxIterations,
          explorationConstant: input.explorationConstant,
          maxDepth: input.maxDepth,
          outputPath: input.outputPath
        }
      });
      logger.info('MCTS process started');
    }
  } catch (error) {
    // Create a safe error object without circular references
    const safeError = {
      message: error instanceof Error ? error.message : String(error),
      name: error instanceof Error ? error.name : 'Unknown Error',
      stack: error instanceof Error ? error.stack : undefined
    };
    
    logger.error('Error in controller step', safeError);
    await emit({
      topic: 'review.error',
      data: {
        message: safeError.message,
        timestamp: new Date().toISOString(),
        repository: input.repoUrl,
        outputPath: input.outputPath,
        requirements: input.requirements
      }
    });
  }
};

File: examples/code-review/steps/code-review/controller_step.py
from datetime import datetime
from typing import Dict, List, Optional, Any
from pydantic import BaseModel, Field
import os
import json
from steps.shared.actions import evaluate_commits, Commits, Node, Evaluation

class MCTSControllerInput(BaseModel):
    prompt: str = Field(..., description="The prompt for code review")
    repo_url: str = Field(..., description="URL of the repository to review")
    branch: str = Field(..., description="Branch to review")
    max_iterations: int = Field(default=100, description="Maximum number of MCTS iterations")
    exploration_constant: float = Field(default=1.414, description="Exploration constant for UCB1")
    max_depth: int = Field(default=10, description="Maximum depth of the MCTS tree")
    review_start_commit: Optional[str] = Field(None, description="Starting commit for review")
    review_end_commit: Optional[str] = Field(None, description="Ending commit for review")
    requirements: str = Field(..., description="Requirements for the code review")
    output_path: Optional[str] = Field(None, description="Path to save the review output")

class MCTSControllerState(BaseModel):
    nodes: Dict[str, Node]
    root_id: str
    current_node_id: str
    current_iteration: int
    max_iterations: int
    exploration_constant: float
    max_depth: int
    output_path: Optional[str]

class ErrorData(BaseModel):
    message: str
    timestamp: str
    repository: str
    output_path: Optional[str]
    requirements: str

config = {
    'type': 'event',
    'name': 'MCTSController',
    'description': 'Controls the MCTS process for code review reasoning',
    'subscribes': ['review.requested', 'mcts.backpropagation.completed'],
    'emits': ['mcts.iteration.started', 'mcts.iterations.completed', 'review.error'],
    'flows': ['code-review-flow']
}

async def handler(req: MCTSControllerInput, context: Any):
    """Handle the MCTS controller step."""
    # Log input with truncated requirements
    truncated_requirements = (
        f"{req.requirements[:20]}..." 
        if len(req.requirements) > 20 
        else req.requirements
    )
    
    context.logger.info('Analyzing review context', {
        **req.model_dump(),
        'requirements': truncated_requirements,
        'repo_url': req.repo_url
    })

    # Store requirements in state for error handling
    await context.state.set(context.trace_id, 'requirements', req.requirements)
 
    try:
        context.logger.info('Creating commits object for repository', { 
            'repo_url': req.repo_url, 
            'branch': req.branch,
            'review_start_commit': req.review_start_commit,
            'review_end_commit': req.review_end_commit
        })
        
        commits = await Commits.create(context.trace_id, context.state, req.model_dump())
        
        context.logger.info('Successfully accessed repository and loaded commits', {
            'files_changed': len(commits.files.split('\n')),
            'commit_messages': len(commits.messages.split('\n'))
        })
        
        evaluation = await evaluate_commits(commits, req.requirements)

        # Define a unique root node ID
        root_id = f'root-{int(datetime.now().timestamp())}'

        # Initialize the MCTS tree structure
        nodes = {
            root_id: Node(
                id=root_id,
                parent=None,
                children=[],
                visits=1,
                value=0,
                state=evaluation.summary,
                isTerminal=False
            )
        }
    
        state = MCTSControllerState(
            nodes=nodes,
            root_id=root_id,
            current_node_id=root_id,
            current_iteration=0,
            max_iterations=req.max_iterations,
            exploration_constant=req.exploration_constant,
            max_depth=req.max_depth,
            output_path=req.output_path
        )
    
        if evaluation.score > 0.9 or req.max_iterations == 0:
            # If the score is already high or no iterations requested, complete immediately
            await context.emit({
                'topic': 'mcts.iterations.completed',
                'data': state.model_dump()
            })
            context.logger.info('Context analysis completed without iterations')
        else:
            # Start the MCTS process
            await context.emit({
                'topic': 'mcts.iteration.started',
                'data': state.model_dump()
            })
            context.logger.info('MCTS process started')
            
    except Exception as error:
        # Create a safe error object without circular references
        error_data = ErrorData(
            message=str(error),
            timestamp=datetime.now().isoformat(),
            repository=req.repo_url,
            output_path=req.output_path,
            requirements=req.requirements
        )
        
        context.logger.error('Error in controller step', {
            'error': str(error),
            'type': type(error).__name__,
            'trace': getattr(error, '__traceback__', None)
        })
        
        await context.emit({
            'topic': 'review.error',
            'data': error_data.model_dump()
        }) 